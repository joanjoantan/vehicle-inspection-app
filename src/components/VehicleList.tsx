import React from "react";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TableRow from "@mui/material/TableRow";
import Paper from "@mui/material/Paper";
import Box from "@mui/material/Box";

import { VehicleListProps } from "./VehicleListInterface";

const VehicleList: React.FC<VehicleListProps> = ({
  vehicles,
  inspections,
  onAttachInspection,
  onRemoveInspection,
  filters,
}) => {
  const filteredVehicles = vehicles.filter((vehicle) => {
    const { vehicleName, vehicleType, allowedType } = filters;

    const matchesName = vehicle.description
      .toLowerCase()
      .includes(vehicleName.toLowerCase());
    const matchesType = vehicleType === "" || vehicle.type === vehicleType;
    const hasMatchingInspection = inspections.some((inspection) => {
      return (
        vehicle.inspections.includes(inspection.id) &&
        (allowedType === "" ||
          inspection.allowedTypes.includes(allowedType) ||
          inspection.allowedTypes.includes("any"))
      );
    });
    const matchesRegistration =
      vehicleName === "" ||
      vehicle.description.toLowerCase().includes(vehicleName.toLowerCase());

    return (
      matchesName && matchesType && hasMatchingInspection && matchesRegistration
    );
  });

  return (
    <Box className="box-container">
      <h2 className="primary-colour">List of Vehicles</h2>
      <TableContainer component={Paper}>
        <Table sx={{ minWidth: 650 }} size="small" aria-label="table">
          <TableHead>
            <TableRow>
              <TableCell>Description</TableCell>
              <TableCell>Type</TableCell>
              <TableCell>Inspections</TableCell>
              <TableCell>Attach Inspections to Vehicles</TableCell>
            </TableRow>
          </TableHead>

          <TableBody>
            {filteredVehicles.map((vehicle) => (
              <TableRow key={vehicle.id}>
                <TableCell> {vehicle.description} </TableCell>

                <TableCell>{vehicle.type}</TableCell>

                <TableCell>
                  {vehicle.inspections.map((inspectionId) => {
                    const inspection = inspections.find(
                      (i) => i.id === inspectionId
                    );
                    if (inspection) {
                      return (
                        <span key={inspection.id}>
                          {inspection.name}{" "}
                          <button
                            onClick={() =>
                              onRemoveInspection(vehicle.id, inspection.id)
                            }
                          >
                            Remove
                          </button>
                          <br />
                        </span>
                      );
                    }
                    return null;
                  })}
                </TableCell>

                <TableCell>
                  <select
                    data-testid={`attach-inspection-select-${vehicle.id}`}
                    onChange={(e) =>
                      onAttachInspection(vehicle.id, parseInt(e.target.value))
                    }
                  >
                    <option value="">Select an inspection</option>
                    {inspections.map((inspection) => (
                      <option
                        key={inspection.id}
                        value={inspection.id}
                        disabled={
                          !inspection.allowedTypes.includes(vehicle.type) &&
                          !inspection.allowedTypes.includes("any")
                        }
                      >
                        {inspection.name}
                      </option>
                    ))}
                  </select>
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
    </Box>
  );
};

export default VehicleList;
