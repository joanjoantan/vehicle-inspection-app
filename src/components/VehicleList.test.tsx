import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import VehicleList from "./VehicleList";

// Mock data for vehicles and inspections
const mockVehicles = [
  { id: 1, description: "Car 1", type: "Sedan", inspections: [1, 2] },
  { id: 2, description: "Van 1", type: "Minivan", inspections: [2] },
  // Add more mock vehicles as needed
];

const mockInspections = [
  { id: 1, name: "Inspection 1", allowedTypes: ["Sedan", "Minivan"] },
  { id: 2, name: "Inspection 2", allowedTypes: ["Minivan"] },
  // Add more mock inspections as needed
];

test("VehicleList renders correctly", () => {
  const onAttachInspection = jest.fn();
  const onRemoveInspection = jest.fn();

  render(
    <VehicleList
      vehicles={mockVehicles}
      inspections={mockInspections}
      onAttachInspection={onAttachInspection}
      onRemoveInspection={onRemoveInspection}
      filters={{
        vehicleName: "",
        vehicleType: "",
        allowedType: "",
        inspectionName: "",
      }}
    />
  );

  // Check if the title is rendered
  expect(screen.getByText("List of Vehicles")).toBeInTheDocument();

  // Check if the table headers are rendered
  expect(screen.getByText("Description")).toBeInTheDocument();
  expect(screen.getByText("Type")).toBeInTheDocument();
  expect(screen.getByText("Inspections")).toBeInTheDocument();
  expect(
    screen.getByText("Attach Inspections to Vehicles")
  ).toBeInTheDocument();

  // Check if vehicle data is displayed
  for (const vehicle of mockVehicles) {
    expect(screen.getByText(vehicle.description)).toBeInTheDocument();
    expect(screen.getByText(vehicle.type)).toBeInTheDocument();
  }

  // Simulate attaching an inspection
  const selectElement = screen.getByTestId("attach-inspection-select-1");
  fireEvent.change(selectElement, { target: { value: "1" } });
  expect(onAttachInspection).toHaveBeenCalledWith(1, 1);
});

test("VehicleList snapshots to compare the rendered component to a saved snapshot", () => {
  const onAttachInspection = jest.fn();
  const onRemoveInspection = jest.fn();

  const { asFragment } = render(
    <VehicleList
      vehicles={mockVehicles}
      inspections={mockInspections}
      onAttachInspection={onAttachInspection}
      onRemoveInspection={onRemoveInspection}
      filters={{
        vehicleName: "",
        vehicleType: "",
        allowedType: "",
        inspectionName: "",
      }}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});
