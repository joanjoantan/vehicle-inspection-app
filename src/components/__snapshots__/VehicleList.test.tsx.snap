// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InspectionList snapshots to compare the rendered component to a saved snapshot 1`] = `
<DocumentFragment>
  <div
    class="box-container MuiBox-root css-0"
  >
    <h2
      class="primary-colour"
    >
      List of Vehicles
    </h2>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="table"
        class="MuiTable-root css-quj9j6-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Description
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Type
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Inspections
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Attach Inspections to Vehicles
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
               Car 1 
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              Sedan
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              <span>
                Inspection 1 
                <button>
                  Remove
                </button>
                <br />
              </span>
              <span>
                Inspection 2 
                <button>
                  Remove
                </button>
                <br />
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              <select
                data-testid="attach-inspection-select-1"
              >
                <option
                  value=""
                >
                  Select an inspection
                </option>
                <option
                  value="1"
                >
                  Inspection 1
                </option>
                <option
                  disabled=""
                  value="2"
                >
                  Inspection 2
                </option>
              </select>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
               Van 1 
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              Minivan
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              <span>
                Inspection 2 
                <button>
                  Remove
                </button>
                <br />
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              <select
                data-testid="attach-inspection-select-2"
              >
                <option
                  value=""
                >
                  Select an inspection
                </option>
                <option
                  value="1"
                >
                  Inspection 1
                </option>
                <option
                  value="2"
                >
                  Inspection 2
                </option>
              </select>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`VehicleList snapshots to compare the rendered component to a saved snapshot 1`] = `
<DocumentFragment>
  <div
    class="box-container MuiBox-root css-0"
  >
    <h2
      class="primary-colour"
    >
      List of Vehicles
    </h2>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="table"
        class="MuiTable-root css-quj9j6-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Description
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Type
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Inspections
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Attach Inspections to Vehicles
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
               Car 1 
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              Sedan
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              <span>
                Inspection 1 
                <button>
                  Remove
                </button>
                <br />
              </span>
              <span>
                Inspection 2 
                <button>
                  Remove
                </button>
                <br />
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              <select
                data-testid="attach-inspection-select-1"
              >
                <option
                  value=""
                >
                  Select an inspection
                </option>
                <option
                  value="1"
                >
                  Inspection 1
                </option>
                <option
                  disabled=""
                  value="2"
                >
                  Inspection 2
                </option>
              </select>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
               Van 1 
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              Minivan
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              <span>
                Inspection 2 
                <button>
                  Remove
                </button>
                <br />
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              <select
                data-testid="attach-inspection-select-2"
              >
                <option
                  value=""
                >
                  Select an inspection
                </option>
                <option
                  value="1"
                >
                  Inspection 1
                </option>
                <option
                  value="2"
                >
                  Inspection 2
                </option>
              </select>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
